pkg/sdk/validation/typeinstance.go

```diff
@@ -0,0 +1,143 @@
+)
+
+type typeInstanceData struct {
+	alias               *string
+	id                  string
+	value               interface{}
+	typeRefWithRevision string
+}
+
+// TypeInstanceValidationHubClient defines Hub methods needed for validation of TypeInstances.
+type TypeInstanceValidationHubClient interface {
+	HubClient
+	FindTypeInstancesTypeRef(ctx context.Context, ids []string) (map[string]graphqllocal.TypeInstanceTypeReference, error)
+}
+
+// ValidateTypeInstancesToCreate  is responsible for validating TypeInstance which do not exist and will be created.
+func ValidateTypeInstancesToCreate(ctx context.Context, client TypeInstanceValidationHubClient, typeInstance *graphqllocal.CreateTypeInstancesInput) (Result, error) {
+	var typeInstanceCollection []*typeInstanceData
+	typeRefCollection := TypeRefCollection{}
+
+	for _, ti := range typeInstance.TypeInstances {
+		if ti == nil {
+			continue
+		}
+		typeRef := types.TypeRef{
+			Path:     ti.TypeRef.Path,
+			Revision: ti.TypeRef.Revision,
+		}
+		name := getManifestPathWithRevision(ti.TypeRef.Path, ti.TypeRef.Revision)
+		typeRefCollection[name] = TypeRef{
+			TypeRef: typeRef,
+		}
+		typeInstanceCollection = append(typeInstanceCollection, &typeInstanceData{
+			alias:               ti.Alias,
+			value:               ti.Value,
+			typeRefWithRevision: name,
+		})
+	}
+
+	schemasCollection, err := ResolveTypeRefsToJSONSchemas(ctx, client, typeRefCollection)
+	if err != nil {
+		return nil, errors.Wrapf(err, "while resolving TypeRefs to JSON Schemas")
+	}
+	return validateTypeInstances(schemasCollection, typeInstanceCollection)
+}
+
+// ValidateTypeInstanceToUpdate  is responsible for validating TypeInstance which exists and will be updated.
+func ValidateTypeInstanceToUpdate(ctx context.Context, client TypeInstanceValidationHubClient, typeInstanceToUpdate []graphqllocal.UpdateTypeInstancesInput) (Result, error) {
+	var typeInstanceIds []string
+	idToTypeNameMap := map[string]string{}
+	for _, ti := range typeInstanceToUpdate {
+		typeInstanceIds = append(typeInstanceIds, ti.ID)
+	}
+
+	typeInstancesTypeRef, err := client.FindTypeInstancesTypeRef(ctx, typeInstanceIds)
+	if err != nil {
+		return nil, errors.Wrapf(err, "while finding TypeInstance Type reference")
+	}
+
+	typeRefCollection := TypeRefCollection{}
+	for id, typeReference := range typeInstancesTypeRef {
+		name := getManifestPathWithRevision(typeReference.Path, typeReference.Revision)
+		typeRefCollection[name] = TypeRef{
+			TypeRef: types.TypeRef{
+				Path:     typeReference.Path,
+				Revision: typeReference.Revision,
+			},
+		}
+		idToTypeNameMap[id] = name
+	}
+
+	var typeInstanceCollection []*typeInstanceData
+	for _, ti := range typeInstanceToUpdate {
+		if ti.TypeInstance == nil {
+			continue
+		}
+		typeInstanceCollection = append(typeInstanceCollection, &typeInstanceData{
+			id:                  ti.ID,
+			value:               ti.TypeInstance.Value,
+			typeRefWithRevision: idToTypeNameMap[ti.ID],
```

Comment:
Why do you need the `idToTypeNameMap`, as you already have the data in `typeInstancesTypeRef`?
If you apply my suggestions regarding data structure, that would be:

```go
	        typeRef, ok := typeInstancesTypeRef[ti.ID]
		if !ok {
			// return error
		}
		
		typeInstanceCollection = append(typeInstanceCollection, &TypeInstanceEssentialData{
			ID:    ptr.String(ti.ID),
			Value: ti.TypeInstance.Value,
			typeRef: types.ManifestRef{
				Path:     typeRef.Path,
				Revision: typeRef.Revision,
			},
		})
```

And you don't need `idToTypeNameMap` at all.